Indicators/ImpliedVolatility.cs

```diff
@@ -0,0 +1,290 @@
+using QuantConnect.Logging;
+using QuantConnect.Python;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// Implied Volatility indicator that calculate the IV of an option using Black-Scholes Model
+    /// </summary>
+    public class ImpliedVolatility : IndicatorBase<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        private readonly Symbol _optionSymbol;
+        private readonly Symbol _underlyingSymbol;
+        private BaseDataConsolidator _consolidator;
+        private RateOfChange _roc;
+        private decimal _impliedVolatility;
+        private OptionPricingModelType _optionModel;
+
+        /// <summary>
+        /// Risk-free rate model
+        /// </summary>
+        private readonly IRiskFreeInterestRateModel _riskFreeInterestRateModel;
+
+        /// <summary>
+        /// Gets the expiration time of the option
+        /// </summary>
+        public DateTime Expiry => _optionSymbol.ID.Date;
+
+        /// <summary>
+        /// Gets the option right (call/put) of the option
+        /// </summary>
+        public OptionRight Right => _optionSymbol.ID.OptionRight;
+
+        /// <summary>
+        /// Gets the strike price of the option
+        /// </summary>
+        public decimal Strike => _optionSymbol.ID.StrikePrice;
+
+        /// <summary>
+        /// Gets the option style (European/American) of the option
+        /// </summary>
+        public OptionStyle Style => _optionSymbol.ID.OptionStyle;
+
+        /// <summary>
+        /// Risk Free Rate
+        /// </summary>
+        public Identity RiskFreeRate { get; set; }
+
+        /// <summary>
+        /// Gets the historical volatility of the underlying
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> HistoricalVolatility { get; }
+
+        /// <summary>
+        /// Gets the option price level
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> Price { get; }
+
+        /// <summary>
+        /// Gets the underlying's price level
+        /// </summary>
+        public IndicatorBase<IndicatorDataPoint> UnderlyingPrice { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the ImpliedVolatility class
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="option">The option to be tracked</param>
+        /// <param name="riskFreeRateModel">Risk-free rate model</param>
+        /// <param name="period">The lookback period of historical volatility</param>
+        /// <param name="optionModel">The option pricing model used to estimate IV</param>
+        public ImpliedVolatility(string name, Symbol option, IRiskFreeInterestRateModel riskFreeRateModel, int period = 252, 
+            OptionPricingModelType optionModel = OptionPricingModelType.BlackScholes)
+            : base(name)
+        {
+            var sid = option.ID;
+            if (sid.SecurityType != SecurityType.Option)
```

Comment:
Oh just realized this should be `if (!sid.SecurityType.IsOption())`, to include FOPs and index options